cmd/juju/cloud/show.go

```diff
@@ -103,56 +110,131 @@ func (c *showCloudCommand) Init(args []string) error {
 	default:
 		return errors.New("no cloud specified")
 	}
-	var err error
-	c.ControllerName, err = c.ControllerNameFromArg()
-	if err != nil {
-		return errors.Wrap(err, errors.New(err.Error()+"\nUse --client-only to query this client."))
-	}
 	return cmd.CheckEmpty(args[1:])
 }
 
 func (c *showCloudCommand) Info() *cmd.Info {
 	return jujucmd.Info(&cmd.Info{
 		Name:    "show-cloud",
 		Args:    "<cloud name>",
-		Purpose: "Shows detailed information on a cloud.",
+		Purpose: "Shows detailed information for a cloud.",
 		Doc:     showCloudDoc,
 	})
 }
 
 func (c *showCloudCommand) Run(ctxt *cmd.Context) error {
 	var (
-		cloud *CloudDetails
-		err   error
+		localCloud *CloudDetails
+		localErr   error
 	)
-	if c.ControllerName == "" {
-		cloud, err = c.getLocalCloud()
-	} else {
-		cloud, err = c.getControllerCloud()
+	if c.BothClientAndController || c.ClientOnly {
+		if localCloud, localErr = c.getLocalCloud(); c.ClientOnly && localErr != nil {
+			return localErr
+		}
 	}
-	if err != nil {
-		return err
+
+	var (
+		remoteCloud *CloudDetails
+		remoteErr   error
+	)
+	if c.BothClientAndController || c.ControllerOnly {
+		if c.ControllerName == "" {
+			// The user may have specified the controller via a --controller option.
+			// If not, let's see if there is a current controller that can be detected.
+			c.ControllerName, remoteErr = c.MaybePromptCurrentController(ctxt, fmt.Sprintf("show cloud %q from", c.CloudName))
+		}
+		if c.ControllerName == "" && remoteErr == nil {
+			remoteErr = errors.New("Not showing a cloud from a controller: no controller specified.")
+		}
+		if remoteErr == nil {
+			remoteCloud, remoteErr = c.getControllerCloud()
+		}
 	}
 
-	displayCloud := cloud
-	displayCloud.CloudType = displayCloudType(displayCloud.CloudType)
-	if err := c.out.Write(ctxt, displayCloud); err != nil {
+	var displayErr error
+	showRemoteConfig := c.includeConfig
+	if localCloud != nil && remoteCloud != nil {
+		// It's possible that a local cloud named A is different to
+		// a remote cloud named A. If their types are different and we
+		// need to display config, we'd need to list each cloud type config.
+		// If the clouds' types are the same, we only need to list
+		// config once after the local cloud information.
+		showRemoteConfig = showRemoteConfig && localCloud.CloudType != remoteCloud.CloudType
+	}
+	if c.BothClientAndController || c.ControllerOnly {
+		if remoteCloud != nil {
+			if err := c.displayCloud(ctxt, remoteCloud, fmt.Sprintf("Cloud %q from controller %q:\n", c.CloudName, c.ControllerName), showRemoteConfig, remoteErr); err != nil {
+				ctxt.Warningf("%v", err)
+				displayErr = cmd.ErrSilent
+			}
+		} else {
+			if remoteErr != nil {
+				ctxt.Warningf("%v", remoteErr)
+				displayErr = cmd.ErrSilent
+			} else {
+				ctxt.Infof("No cloud %q exists remotely on the controller.", c.CloudName)
```

Comment:
do we need the word "remotely" here. I think "exists on the controller" is sufficient and clearer